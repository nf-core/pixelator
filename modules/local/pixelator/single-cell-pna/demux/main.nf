process PIXELATOR_PNA_DEMUX {
    tag "${meta.id}"
    label 'process_medium'

    // TODO: Add conda
    // conda "bioconda::pixelator=0.18.2"

    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'quay.io/pixelgen-technologies/pixelator:0.21.2'
        : 'quay.io/pixelgen-technologies/pixelator:0.21.2'}"

    input:
    tuple val(meta), path(reads), path(panel_file), val(panel), val(design)

    output:
    tuple val(meta), path("demux/*.parquet", arity: "1..*"),      emit: demuxed
    tuple val(meta), path("demux/*demux.passed*.{fq,fastq}.zst"), emit: passed
    tuple val(meta), path("demux/*demux.failed.{fq,fastq}.zst"),  emit: failed
    tuple val(meta), path("demux/*.report.json"),                 emit: report_json
    tuple val(meta), path("demux/*.meta.json"),                   emit: metadata_json
    tuple val(meta), path("*pixelator-demux.log"),                emit: log

    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    // --design is passed in meta and added to args through modules.conf

    prefix = task.ext.prefix ?: "${meta.id}"
    def args = task.ext.args ?: ''
    def panelOpt = (panel
        ? "--panel ${panel}"
        : panel_file ? "--panel ${panel_file}" : "")
    def designOpt = "--design ${design}"

    """
    pixelator \\
        --log-file ${prefix}.pixelator-demux.log \\
        --verbose \\
        single-cell-pna \\
        demux \\
        --threads ${task.cpus} \\
        --output . \\
        ${panelOpt} \\
        ${designOpt} \\
        ${args} \\
        ${reads}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """

    stub:
    prefix = task.ext.prefix ?: "${meta.id}"

    """
    mkdir demux
    touch demux/${prefix}.report.json
    touch demux/${prefix}.meta.json
    touch demux/${prefix}.demux.passed.fq.zst
    touch demux/${prefix}.demux.failed.fq.zst
    touch demux/${prefix}.demux.m1.part_000.parquet
    touch demux/${prefix}.demux.m2.part_000.parquet
    touch ${prefix}.pixelator-demux.log


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """
}
