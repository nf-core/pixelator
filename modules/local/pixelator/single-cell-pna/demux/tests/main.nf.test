nextflow_process {

    name "Test Process PIXELATOR_PNA_DEMUX"
    script "../main.nf"
    process "PIXELATOR_PNA_DEMUX"
    tag "modules"
    tag "pixelator"
    tag "pixelator/pna"
    tag "pixelator/single_cell_pna_demux"

    test("PNA demux - small test") {
        when {
            process {
                """
                input[0] = [
                    [ id:'PNA055_Sample07_filtered_S7', design:'pna-2', panel:'proxiome-immuno-155', technology:'pna' ],
                    file(params.pipelines_testdata_base_path + 'testdata/pna/modules/amplicon/PNA055_Sample07_filtered_S7.amplicon.fq.zst', checkIfExists: true),
                    [],
                    'proxiome-immuno-155',
                    'pna-2'
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                {
                    with(process.out.demuxed) {
                        assert get(0).get(1).each { file(it).exists() }
                    }
                },
                { assert snapshot(process.out.passed).match("passed") },
                { assert snapshot(process.out.failed).match("failed") },
                { assert snapshot(process.out.report_json).match("report_json") },
                { assert snapshot(process.out.metadata_json).match("metadata_json") },
                {
                    with (process.out.log) {
                        assert path(get(0).get(1)).readLines().last().contains("Finished pixelator demux")
                    }
                },
            )
        }
    }

    test("PNA demux - multiple chunks") {
        config "./multiple_chunks.config"

        when {
            process {
                """
                input[0] = [
                    [ id:'PNA055_Sample07_filtered_S7', design:'pna-2', panel:'proxiome-immuno-155', technology:'pna' ],
                    file(params.pipelines_testdata_base_path + 'testdata/pna/modules/amplicon/PNA055_Sample07_filtered_S7.amplicon.fq.zst', checkIfExists: true),
                    [],
                    'proxiome-immuno-155',
                    'pna-2'
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                {
                    with(process.out.demuxed) {
                        assert get(0).get(1).each { file(it).exists() }
                    }
                },
                { assert snapshot(process.out.passed).match("multi_passed") },
                { assert snapshot(process.out.failed).match("multi_failed") },
                { assert snapshot(process.out.report_json).match("multi_report_json") },
                { assert snapshot(process.out.metadata_json).match("multi_metadata_json") },
                {
                    with (process.out.log) {
                        assert path(get(0).get(1)).readLines().last().contains("Finished pixelator demux")
                    }
                },
            )
        }
    }
}

