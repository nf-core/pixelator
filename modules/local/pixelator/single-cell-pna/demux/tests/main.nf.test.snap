{
    "multi_passed": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.demux.passed.fq.zst:md5,22d1ba0788bc719ff5086df10efc79c0"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-20T10:45:19.870166753"
    },
    "report_json": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.report.json:md5,7ebfb9ffcdad63735bf7db85f95b2375"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-20T10:44:33.650948168"
    },
    "multi_failed": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.demux.failed.fq.zst:md5,ec839fc21a49891f10d87bbe122c3896"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-20T10:45:20.092313742"
    },
    "passed": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.demux.passed.fq.zst:md5,22d1ba0788bc719ff5086df10efc79c0"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-20T10:44:33.487122857"
    },
    "failed": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.demux.failed.fq.zst:md5,ec839fc21a49891f10d87bbe122c3896"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-20T10:44:33.568673937"
    },
    "multi_report_json": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.report.json:md5,7ebfb9ffcdad63735bf7db85f95b2375"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-20T10:45:20.283348049"
    },
    "multi_metadata_json": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.meta.json:md5,e264dd4b88e2ef36d8af6640adbe79ac"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-24T17:10:34.346405357"
    },
    "metadata_json": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.meta.json:md5,2bd1ff0d69fc6584e455196d8805f587"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-24T17:09:59.60486159"
    }
}